src/pyclass.rs

```diff
@@ -107,134 +138,173 @@ pub trait PyClass:
-            p.as_ptr()
-        })
-        .unwrap_or_else(|| nb_bool.map_or_else(ptr::null_mut, ffi::PyNumberMethods::from_nb_bool));
+    if let Some(number) = T::number_methods() {
+        unsafe { number.as_ref() }.update_slots(&mut slots);
+    }
+    // iterator methods
+    if let Some(iter) = T::iter_methods() {
+        unsafe { iter.as_ref() }.update_slots(&mut slots);
+    }
     // mapping methods
-    type_object.tp_as_mapping = T::mapping_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
+    if let Some(mapping) = T::mapping_methods() {
+        unsafe { mapping.as_ref() }.update_slots(&mut slots);
+    }
     // sequence methods
-    type_object.tp_as_sequence = T::sequence_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
+    if let Some(sequence) = T::sequence_methods() {
+        unsafe { sequence.as_ref() }.update_slots(&mut slots);
+    }
+    // descriptor protocol
+    if let Some(descr) = T::descr_methods() {
+        unsafe { descr.as_ref() }.update_slots(&mut slots);
+    }
     // async methods
-    type_object.tp_as_async = T::async_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
-    // buffer protocol
-    type_object.tp_as_buffer = T::buffer_methods().map_or_else(ptr::null_mut, |p| p.as_ptr());
+    if let Some(asnc) = T::async_methods() {
+        unsafe { asnc.as_ref() }.update_slots(&mut slots);
+    }
+    // gc methods
+    if let Some(gc) = T::gc_methods() {
+        unsafe { gc.as_ref() }.update_slots(&mut slots);
+    }
 
-    let (new, call, mut methods) = py_class_method_defs::<T>();
+    slots.push(0, ptr::null_mut());
+    let mut spec = ffi::PyType_Spec {
+        name: get_type_name::<T>(module_name)?,
+        basicsize: std::mem::size_of::<T::Layout>() as c_int,
+        itemsize: 0,
+        flags: py_class_flags::<T>(),
+        slots: slots.0.as_mut_slice().as_mut_ptr(),
+    };
 
-    // normal methods
-    if !methods.is_empty() {
-        methods.push(ffi::PyMethodDef_INIT);
-        type_object.tp_methods = Box::into_raw(methods.into_boxed_slice()) as _;
+    let type_object = unsafe { ffi::PyType_FromSpec(&mut spec) };
+    if type_object.is_null() {
+        Err(PyErr::fetch(py))
+    } else {
+        tp_init_additional::<T>(type_object as _);
+        Ok(type_object as _)
     }
+}
 
-    // __new__ method
-    type_object.tp_new = new;
-    // __call__ method
-    type_object.tp_call = call;
-
-    // properties
-    let mut props = py_class_properties::<T>();
+#[cfg(not(Py_LIMITED_API))]
+fn tp_init_additional<T: PyClass>(type_object: *mut ffi::PyTypeObject) {
+    // Just patch the type objects for the things there's no
+    // PyType_FromSpec API for... there's no reason this should work,
+    // except for that it does and we have tests.
+
+    // Running this causes PyPy to segfault.
+    #[cfg(not(PyPy))]
+    if T::DESCRIPTION != "\0" {
+        unsafe {
+            // Until CPython 3.10, tp_doc was treated specially for heap-types,
```

Comment:
Do we need to add some kind of `#[cfg(Py_3_10)]` branch to this code?

Reply:
Good idea.